cadence/server/api_actions.go

```diff
@@ -0,0 +1,197 @@
+// api_actions.go
+// Repeatable actions used by API functions. Mostly database and audio software calls.
+// No direct responses to clients here.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/Jeffail/gabs"
+	"github.com/kenellorando/clog"
+)
+
+// Song file metadata object
+type SongData struct {
+	ID     int
+	Artist string
+	Title  string
+	Album  string
+	Genre  string
+	Year   int
+}
+
+// Takes a query string to search the database.
+// Returns a slice of SongData of songs ordered by relevance.
+func searchByQuery(query string) (queryResults []SongData, err error) {
+	clog.Info("searchByQuery", fmt.Sprintf("Searching database for query: '%v'", query))
+
+	selectWhereStatement := fmt.Sprintf("SELECT \"rowid\", \"artist\", \"title\",\"album\", \"genre\", \"year\" FROM %s ", c.MetadataTable) + "WHERE artist LIKE $1 OR title LIKE $2 ORDER BY rank"
+	rows, err := db.Query(selectWhereStatement, "%"+query+"%", "%"+query+"%")
+	if err != nil {
+		clog.Error("searchByQuery", "Database search failed.", err)
+		return nil, err
+	}
+
+	for rows.Next() {
+		song := &SongData{}
+		err = rows.Scan(&song.ID, &song.Artist, &song.Title, &song.Album, &song.Genre, &song.Year)
+		if err != nil {
+			clog.Error("searchByQuery", "Data scan failed.", err)
+			continue
+		}
+		queryResults = append(queryResults, SongData{ID: song.ID, Artist: song.Artist, Title: song.Title, Album: song.Album, Genre: song.Genre, Year: song.Year})
+	}
+
+	return queryResults, nil
+}
+
+// Takes a title and artist string to find a song which exactly matches.
+// Returns a slice of SongData of songs by relevance.
+// This search should only have one result unless multiple audio files share the exact same title and artist.
+func searchByTitleArtist(title string, artist string) (queryResults []SongData, err error) {
+	selectStatement := fmt.Sprintf("SELECT rowid,artist,title,album,genre,year FROM %s WHERE title=\"%v\" AND artist=\"%v\";", c.MetadataTable, title, artist)
+	rows, err := db.Query(selectStatement)
+	if err != nil {
+		clog.Error("searchByTitleArtist", "Could not query DB.", err)
+		return nil, err
+	}
+
+	for rows.Next() {
+		song := &SongData{}
+		err = rows.Scan(&song.ID, &song.Artist, &song.Title, &song.Album, &song.Genre, &song.Year)
+		if err != nil {
+			clog.Error("searchByTitleArtist", "Data scan failed.", err)
+			continue
+		}
+		queryResults = append(queryResults, SongData{ID: song.ID, Artist: song.Artist, Title: song.Title, Album: song.Album, Genre: song.Genre, Year: song.Year})
+	}
+
+	return queryResults, nil
+}
+
```

Comment:
No support for things like the genre search, I take it?

Reply:
Nope. I decided to stick strictly to title and artist because that is the only information available (that I want to display) on the frontend. It might be confusing for some results to appear or rank higher when the partial search term for one thing exactly matches another.